src/callback_signer.rs

```diff
@@ -0,0 +1,351 @@
+// Copyright 2024 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License,
+// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+// or the MIT license (http://opensource.org/licenses/MIT),
+// at your option.
+// Unless required by applicable law or agreed to in writing,
+// this software is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
+// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
+// specific language governing permissions and limitations under
+// each license.
+
+use std::{
+    io::Write,
+    path::PathBuf,
+    process::{Command, Stdio},
+};
+
+use anyhow::{bail, Context};
+use c2pa::{Error, Signer, SigningAlg};
+
+use crate::signer::SignConfig;
+
+/// A struct that implements [SignCallback]. This struct will call out to the client provided
+/// external signer to get the signed bytes for the asset.
+pub(crate) struct ExternalProcessRunner {
+    config: CallbackSignerConfig,
+    process_name: String,
+}
+
+impl ExternalProcessRunner {
+    pub fn new(config: CallbackSignerConfig, process_name: String) -> Self {
+        Self {
+            config,
+            process_name,
+        }
+    }
+}
+
+impl SignCallback for ExternalProcessRunner {
+    /// Runs the client-provided [Command], passing to it, via stdin, the bytes to be signed. We
+    /// also pass the `reserve-size`, `sign-cert`, and `alg` as CLI arguments to the [Command].
+    fn sign(&self, bytes: &[u8]) -> anyhow::Result<Vec<u8>> {
+        let sign_cert = self
+            .config
+            .sign_cert_path
+            .as_os_str()
+            .to_str()
+            .context("Unable to read sign_certs. Is the sign_cert path valid?")?;
+
+        // Spawn external process provided by the `c2patool` client.
+        let mut child = Command::new(&self.process_name)
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
```

Comment:
do we need to add stderr()  here?

Reply:
Yes, you're right. Will add 👍

Reply:
Resolved here: https://github.com/contentauth/c2patool/pull/169/commits/86bebfc451e75be1addf406c564748aeb5905616